src/v/archival/ntp_archiver_service.cc

```diff
@@ -293,6 +302,11 @@ ss::future<> ntp_archiver::upload_topic_manifest() {
 ss::future<> ntp_archiver::upload_until_term_change() {
     ss::lowres_clock::duration backoff = _conf->upload_loop_initial_backoff;
 
+    // Hold sempahore units to enable other code to know that we are in
+    // the process of doing uploads + wait for us to drop out if they
+    // e.g. set _paused.
+    auto units = co_await ss::get_units(_uploads_active, 1);
```

Comment:
This can cause a deadlock because we're taking this mutex in `upload_next_candidates` and the mutex is not recursive.